### PR TITLE
Add new features to schema inference for JSON formats

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This PRs improves schema inference from JSON formats:
1) Now it's possible to infer named Tuples from JSON objects without experimantal JSON type under a setting `input_format_json_try_infer_named_tuples_from_objects` in JSON formats. Previously without experimantal type JSON we could only infer JSON objects as Strings or Maps, now we can infer named Tuple. Resulting Tuple type will conain all keys of objects that were read in data sample during schema inference. It can be useful for reading structured JSON data without sparse objects. The setting is enabled by default.
2) Allow parsing JSON array into a column with type String under setting `input_format_json_read_arrays_as_strings`. It can help reading arrays with values with different types.
3) Allow to use type String for JSON keys with unkown types (`null`/`[]`/`{}`) in sample data under setting `input_format_json_infer_incomplete_types_as_strings`. Now in JSON formats we can read any value into String column and we can avoid getting error `Cannot determine type for column 'column_name' by first 25000 rows of data, most likely this column contains only Nulls or empty Arrays/Maps` during schema inference by using type String for unknown types, so the data will be read successfully.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit ba307c74660ef56e00fc55f5f648a090a345e399 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54427/ba307c74660ef56e00fc55f5f648a090a345e399/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by Avogar** [comment]
@alexey-milovidov WHYT about enabling this new behaviour by default? I think it can cover most of the cases with json datasets.
Pros: with more or less structured json data we will infer correct Tuples with all elements in the data, so we won't need JSON type
Cons: with sparse json data resulting Tuples will contain not all elements (only elements we could read during schema inference), so we won't read all object keys.

Right now by default we infer JSON objects as Strings (or Maps if all values in object have the same type) so it can be processed futher by the userwith functions JSONExtract*.

**Comment by alexey-milovidov** [comment]
Yes, let's enable it by default. Looks good for me.

**Comment by alexey-milovidov** [comment]
Only one test remains...

**Comment by alexey-milovidov** [comment]
I've tested it on this file:
[inference_test_1000.jsonl.zst.txt](https://github.com/ClickHouse/ClickHouse/files/12677694/inference_test_1000.jsonl.zst.txt)
and found that it works very slow (it didn't finish, still running).

**Comment by Avogar** [comment]
Example from Alexey witn new optimized impleemntation:
```
avogar-dev :) desc file(inference_test_1000.jsonl)

DESCRIBE TABLE file(inference_test_1000.jsonl)

Query id: 0519e04b-62af-464b-a876-b8b0468a0ab5

┌─name─┬─type───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ ip   │ Nullable(String)                                                                                                                                                                                                                                           │              │                    │         │                  │                │
│ data │ Tuple(http Tuple(error Nullable(String), protocol Nullable(String), result Tuple(redirect_response_chain Array(Tuple(body Nullable(String), body_sha256 Nullable(String), content_length Nullable(Int64), headers Tuple(access_control_allow_origin Array(Nullable(String)), cache_control Array(Nullable(String)), connection Array(Nullable(String)), content_language Array(Nullable(String)), content_length Array(Nullable(Int64)), content_security_policy Array(Nullable(String)), content_type Array(Nullable(String)), date Array(Nullable(String)), expires Array(Nullable(String)), last_modified Array(Nullable(String)), location Array(Nullable(String)), pragma Array(Nullable(String)), server Array(Nullable(String)), set_cookie Array(Nullable(String)), unknown Array(Tuple(key Nullable(String), value Array(Nullable(String)))), upgrade Array(Nullable(String)), vary Array(Nullable(String)), via Array(Nullable(String)), x_frame_options Array(Nullable(String)), x_powered_by Array(Nullable(String)), x_xss_protection Array(Nullable(Int64))), protocol Tuple(major Nullable(Int64), minor Nullable(Int64), name Nullable(String)), request Tuple(headers Tuple(accept Array(Nullable(String)), referer Array(Nullable(String)), user_agent Array(Nullable(String))), host Nullable(String), method Nullable(String), url Tuple(host Nullable(String), path Nullable(String), raw_query Nullable(String), scheme Nullable(String))), status_code Nullable(Int64), status_line Nullable(String), transfer_encoding Array(Nullable(String)))), response Tuple(body Nullable(String), body_sha256 Nullable(String), content_length Nullable(Int64), headers Tuple(accept_ranges Array(Nullable(String)), access_control_allow_origin Array(Nullable(String)), age Array(Nullable(Int64)), alt_svc Array(Nullable(String)), cache_control Array(Nullable(String)), connection Array(Nullable(String)), content_language Array(Nullable(String)), content_length Array(Nullable(Int64)), content_location Array(Nullable(String)), content_security_policy Array(Nullable(String)), content_type Array(Nullable(String)), date Array(Nullable(String)), etag Array(Nullable(String)), expires Array(Nullable(String)), last_modified Array(Nullable(String)), link Array(Nullable(String)), p3p Array(Nullable(String)), pragma Array(Nullable(String)), public_key_pins Array(Nullable(String)), retry_after Array(Nullable(Int64)), server Array(Nullable(String)), set_cookie Array(Nullable(String)), status Array(Nullable(String)), strict_transport_security Array(Nullable(String)), unknown Array(Tuple(key Nullable(String), value Array(Nullable(String)))), upgrade Array(Nullable(String)), vary Array(Nullable(String)), via Array(Nullable(String)), www_authenticate Array(Nullable(String)), x_content_type_options Array(Nullable(String)), x_frame_options Array(Nullable(String)), x_powered_by Array(Nullable(String)), x_ua_compatible Array(Nullable(String)), x_xss_protection Array(Nullable(String))), protocol Tuple(major Nullable(Int64), minor Nullable(Int64), name Nullable(String)), request Tuple(headers Tuple(accept Array(Nullable(String)), referer Array(Nullable(String)), user_agent Array(Nullable(String))), host Nullable(String), method Nullable(String), tls_log Tuple(handshake_log Tuple(client_finished Tuple(verify_data Nullable(String)), client_key_exchange Tuple(ecdh_params Tuple(client_private Tuple(length Nullable(Int64), value Nullable(String)), client_public Tuple(x Tuple(length Nullable(Int64), value Nullable(String)), y Tuple(length Nullable(Int64), value Nullable(String))), curve_id Tuple(id Nullable(Int64), name Nullable(String))), rsa_params Tuple(encrypted_pre_master_secret Nullable(String), length Nullable(Int64))), key_material Tuple(master_secret Tuple(length Nullable(Int64), value Nullable(String)), pre_master_secret Tuple(length Nullable(Int64), value Nullable(String))), server_certificates Tuple(certificate Tuple(parsed Tuple(extensions Tuple(authority_info_access Tuple(issuer_urls Array(Nullable(String)), ocsp_urls Array(Nullable(String))), authority_key_id Nullable(String), basic_constraints Tuple(is_ca Nullable(Bool)), certificate_policies Array(Tuple(cps Array(Nullable(String)), id Nullable(String))), crl_distribution_points Array(Nullable(String)), extended_key_usage Tuple(client_auth Nullable(Bool), server_auth Nullable(Bool)), key_usage Tuple(digital_signature Nullable(Bool), key_encipherment Nullable(Bool), value Nullable(Int64)), signed_certificate_timestamps Array(Tuple(log_id Nullable(String), signature Nullable(String), timestamp Nullable(Int64), version Nullable(Int64))), subject_alt_name Tuple(dns_names Array(Nullable(String))), subject_key_id Nullable(String)), fingerprint_md5 Nullable(String), fingerprint_sha1 Nullable(String), fingerprint_sha256 Nullable(String), issuer Tuple(common_name Array(Nullable(String)), country Array(Nullable(String)), locality Array(Nullable(String)), organization Array(Nullable(String)), organizational_unit Array(Nullable(String)), province Array(Nullable(String))), issuer_dn Nullable(String), names Array(Nullable(String)), redacted Nullable(Bool), serial_number Nullable(Float64), signature Tuple(self_signed Nullable(Bool), signature_algorithm Tuple(name Nullable(String), oid Nullable(String)), valid Nullable(Bool), value Nullable(String)), signature_algorithm Tuple(name Nullable(String), oid Nullable(String)), spki_subject_fingerprint Nullable(String), subject Tuple(common_name Array(Nullable(String)), country Array(Nullable(String)), jurisdiction_country Array(Nullable(String)), locality Array(Nullable(String)), organization Array(Nullable(String)), organizational_unit Array(Nullable(String)), province Array(Nullable(String)), serial_number Array(Nullable(String))), subject_dn Nullable(String), subject_key_info Tuple(ecdsa_public_key Tuple(b Nullable(String), curve Nullable(String), gx Nullable(String), gy Nullable(String), length Nullable(Int64), n Nullable(String), p Nullable(String), pub Nullable(String), x Nullable(String), y Nullable(String)), fingerprint_sha256 Nullable(String), key_algorithm Tuple(name Nullable(String)), rsa_public_key Tuple(exponent Nullable(Int64), length Nullable(Int64), modulus Nullable(String))), tbs_fingerprint Nullable(String), tbs_noct_fingerprint Nullable(String), validation_level Nullable(String), validity Tuple(end Nullable(String), length Nullable(Int64), start Nullable(String)), version Nullable(Int64)), raw Nullable(String)), chain Array(Tuple(parsed Tuple(extensions Tuple(authority_info_access Tuple(issuer_urls Array(Nullable(String)), ocsp_urls Array(Nullable(String))), authority_key_id Nullable(String), basic_constraints Tuple(is_ca Nullable(Bool), max_path_len Nullable(Int64)), certificate_policies Array(Tuple(cps Array(Nullable(String)), id Nullable(String))), crl_distribution_points Array(Nullable(String)), extended_key_usage Tuple(client_auth Nullable(Bool), server_auth Nullable(Bool)), key_usage Tuple(certificate_sign Nullable(Bool), crl_sign Nullable(Bool), digital_signature Nullable(Bool), value Nullable(Int64)), subject_key_id Nullable(String)), fingerprint_md5 Nullable(String), fingerprint_sha1 Nullable(String), fingerprint_sha256 Nullable(String), issuer Tuple(common_name Array(Nullable(String)), country Array(Nullable(String)), locality Array(Nullable(String)), organization Array(Nullable(String)), organizational_unit Array(Nullable(String)), province Array(Nullable(String))), issuer_dn Nullable(String), redacted Nullable(Bool), serial_number Nullable(Float64), signature Tuple(self_signed Nullable(Bool), signature_algorithm Tuple(name Nullable(String), oid Nullable(String)), valid Nullable(Bool), value Nullable(String)), signature_algorithm Tuple(name Nullable(String), oid Nullable(String)), spki_subject_fingerprint Nullable(String), subject Tuple(common_name Array(Nullable(String)), country Array(Nullable(String)), locality Array(Nullable(String)), organization Array(Nullable(String)), organizational_unit Array(Nullable(String)), province Array(Nullable(String))), subject_dn Nullable(String), subject_key_info Tuple(fingerprint_sha256 Nullable(String), key_algorithm Tuple(name Nullable(String)), rsa_public_key Tuple(exponent Nullable(Int64), length Nullable(Int64), modulus Nullable(String))), tbs_fingerprint Nullable(String), tbs_noct_fingerprint Nullable(String), validation_level Nullable(String), validity Tuple(end Nullable(String), length Nullable(Int64), start Nullable(String)), version Nullable(Int64)), raw Nullable(String))), validation Tuple(browser_error Nullable(String), browser_trusted Nullable(Bool))), server_finished Tuple(verify_data Nullable(String)), server_hello Tuple(cipher_suite Tuple(hex Nullable(String), name Nullable(String), value Nullable(Int64)), compression_method Nullable(Int64), extended_master_secret Nullable(Bool), heartbeat Nullable(Bool), ocsp_stapling Nullable(Bool), random Nullable(String), secure_renegotiation Nullable(Bool), session_id Nullable(String), ticket Nullable(Bool), version Tuple(name Nullable(String), value Nullable(Int64))), server_key_exchange Tuple(digest Nullable(String), ecdh_params Tuple(curve_id Tuple(id Nullable(Int64), name Nullable(String)), server_public Tuple(x Tuple(length Nullable(Int64), value Nullable(String)), y Tuple(length Nullable(Int64), value Nullable(String)))), signature Tuple(raw Nullable(String), signature_and_hash_type Tuple(hash_algorithm Nullable(String), signature_algorithm Nullable(String)), tls_version Tuple(name Nullable(String), value Nullable(Int64)), type Nullable(String), valid Nullable(Bool))))), url Tuple(host Nullable(String), path Nullable(String), raw_query Nullable(String), scheme Nullable(String))), status_code Nullable(Int64), status_line Nullable(String), transfer_encoding Array(Nullable(String)))), status Nullable(String), timestamp Nullable(String))) │              │                    │         │                  │                │
└──────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

2 rows in set. Elapsed: 0.063 sec.
```

**Comment by Avogar** [comment]
`AST fuzzer (ubsan)` - https://github.com/ClickHouse/ClickHouse/issues/55016

File: docs/en/interfaces/formats.md
Comment (Avogar) [inline docs/en/interfaces/formats.md:1262]:
```suggestion
- [input_format_json_read_numbers_as_strings](/docs/en/operations/settings/settings-formats.md/#input_format_json_read_numbers_as_strings) - allow to parse numbers as strings in JSON input formats. Default value - `true`.
```

File: src/Formats/SchemaInferenceUtils.cpp
Comment (Avogar) [inline src/Formats/SchemaInferenceUtils.cpp:131]:
```suggestion
                        /// If types for the same path are different even after transform, we cannot merge these objects.
```

File: src/Formats/SchemaInferenceUtils.cpp
Comment (Avogar) [inline src/Formats/SchemaInferenceUtils.cpp:329]: I changed logic of experimental JSON type inference. Previously if setting `allow_experimantal_object_type` was enabled, we used `JSON` type for object only if we couldn't infer it as `Map(String, ValueType)`. And I think this logic was not good, no need to use Map type at all, now we always use `JSON` type for objects during inference if `allow_experimantal_object_type` is enabled.

File: src/Formats/SchemaInferenceUtils.cpp
Comment (antonio2368) [inline src/Formats/SchemaInferenceUtils.cpp:1271]: comment seems strange
maybe `Now` at start?

File: src/DataTypes/Serializations/SerializationString.cpp
Comment (Avogar) [reply src/DataTypes/Serializations/SerializationString.cpp:335]: To be honest I just copied the logic from reading Object as String (that was implemented not by me). Of course better to read directly to column data, I will try to change it

File: src/DataTypes/transformTypesRecursively.cpp
Comment (Avogar) [reply src/DataTypes/transformTypesRecursively.cpp:112]: I would say we consider two named tuples equal if they have the same order of elements

File: src/Formats/SchemaInferenceUtils.cpp
Comment (Avogar) [reply src/Formats/SchemaInferenceUtils.cpp:122]: Good idea, will try

Comment (Avogar) [reply src/Formats/SchemaInferenceUtils.cpp]: Done

File: src/DataTypes/Serializations/SerializationString.cpp
Comment (Avogar) [reply src/DataTypes/Serializations/SerializationString.cpp]: Done

**Comment by Avogar** [comment]
@antonio2368 I would appreciate if you could continue reviewing this PR today/tomorrow morning, so I can fix everything and we can merge it before release

File: src/Formats/SchemaInferenceUtils.cpp
Comment (antonio2368) [reply src/Formats/SchemaInferenceUtils.cpp]: sorry, I wrote in my example `insert` and mentioned `try_emplace` because I found it later.
In any case, `try_emplace` seems better because it avoids copying of key and value if it's not needed.
But wait for review, if there won't be other comments we can merge as is.

**Review by antonio2368 — COMMENTED** [review]
Comments are mostly cosmetics, I'll approve after they are applied

File: src/Formats/SchemaInferenceUtils.cpp
Comment (Avogar) [reply src/Formats/SchemaInferenceUtils.cpp]: Forgot to remove it after testing two implmentations. Thanks for noticing, removed

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**
